File: Design-Patterns-Using-Kotlin/Strategy/StrategyPattern.kt
interface Strategy {
    fun execute(data: Int)
}

class ConcreteStrategy1 : Strategy {
    override fun execute(data: Int) {
        println("ConcreteStrategy1 execute")
    }
}

class ConcreteStrategy2 : Strategy {
    override fun execute(data: Int) {
        println("ConcreteStrategy2 execute")
    }
}

class Context(private var strategy: Strategy) {
    fun setStrategy(strategy: Strategy) {
        this.strategy = strategy
    }

    fun execute() {
        val data = 1
        strategy.execute(data)
    }
}

// Client code
fun main() {
    val context = Context(ConcreteStrategy1())
    context.execute()

    context.setStrategy(ConcreteStrategy2())
    context.execute()
}

/*
ConcreteStrategy1 execute
ConcreteStrategy2 execute
*/

File: Design-Patterns-Using-Ruby/Strategy/StrategyPattern.rb
# Define Strategy interface
module Strategy
    def execute(data)
        raise NotImplementedError, 'Subclasses must implement the execute method'
    end
end

# Define ConcreteStrategy1 class implementing Strategy
class ConcreteStrategy1
    include Strategy

    def execute(data)
        puts 'ConcreteStrategy1 execute'
    end
end

# Define ConcreteStrategy2 class implementing Strategy
class ConcreteStrategy2
    include Strategy

    def execute(data)
        puts 'ConcreteStrategy2 execute'
    end
end

# Define Context class
class Context
    attr_accessor :strategy

    def initialize(strategy)
        @strategy = strategy
    end

    def set_strategy(strategy)
        @strategy = strategy
    end

    def execute
        data = 1
        @strategy.execute(data)
    end
end

# Client code
context = Context.new(ConcreteStrategy1.new)
context.execute

context.set_strategy(ConcreteStrategy2.new)
context.execute

=begin 
ConcreteStrategy1 execute
ConcreteStrategy2 execute
=end
